### ISSUE TITLE
Exclude modules from addons from coverage instrumentation

### ISSUE THREAD

**Issue by sglanzer** [sglanzer/ember-cli-blanket#17]
Addon modules should not be part of the coverage report for a project.  Addons should be tested in their own projects and will impact the coverage report of dependent projects if they are not excluded.

**Comment by flyswatter** [comment]
Came here to suggest the same thing.  Is there a plan for how to implement this?

**Comment by sglanzer** [comment]
I haven't really started looking into it - with the new way addons are namespaced in ember-cli 0.2.x it's probably more possible than it was before.  I'd probably start with asking some of the ember-cli core team about this if we want to start investigation.

**Comment by flyswatter** [comment]
Hey @stefpenner, how would you go about excluding ember-addons from build-time tooling like this?

**Comment by flyswatter** [comment]
Hey @stefanpenner, how would you go about excluding ember-addons from build-time tooling like this?

**Comment by stefanpenner** [comment]
maybe some module name pattern matching?

**Comment by jschilli** [comment]
@flyswatter - have you tried the blanket-options.js file?

It's not a perfect answer but I use it in both addons and apps to keep the focus on the code under test:

```

var options = {
  modulePrefix: "my-addon",
  filter: "//.*my-addon/.*/",
  antifilter: "//.*(tests|template).*/",
  loaderExclusions: [],
  enableCoverage: true,
  cliOptions: {
    reporters: ['json']
  }
};
if (typeof exports === 'undefined') {
  blanket.options(options);
} else {
  module.exports = options;
}
```

in one of my apps I set loaderExclusions to:

```
    loaderExclusions: [
    'myapp/modules/domain1/lib',
    'myapp/modules/domain2/lib',
    'myapp/vendor'
    ],
```

the imperfectness comes in from the fact that addons have an addon part (which is present in the resulting app) and the app/* part which is blended by the broccoli build and is indistinguishable from 'normal' app code by the time we annotate it - which is in the browser.

so the standard blanket options can help you narrow the code under cover down pretty significantly

**Comment by flyswatter** [comment]
@jschilli I have used the blanket-options, and like you say, it has some limitations, especially since the trees are merged before the filtering takes place.

I guess very specific exclusions is a solid short-term solution, but long-term it would be better if `ember-cli-blanket` didn't try to run coverage tests on included addons in the first place.

**Comment by jschilli** [comment]
@flyswatter long term would be nice.  In the realistic term we don't have many options based on when and how blanket is being run.

One idea that I've had is to do the blanketing in the plugin in node land to help resolve the sourcemapping discrepancy that is compounded by blanketing in browser.

That said the issue of determining lineage of contributed code other than through pattern mapping will be involved.

I would expect that we won't make progress on this until we re-examine the way in which blanketing is being done and that requires time/capacity.  I know that @sglanzer and I are both maxxed on day jobs ATM so wouldnt' expect this to change in the near term

**Comment by flyswatter** [comment]
Yeah, looking closer at the regex filtering is actually enough for me, I agree this shouldn't be a high priority.

Thanks for your help all!

**Comment by sglanzer** [comment]
I may have some ideas on this - I need to do some prototype work and I'll update afterward.

**Comment by flyswatter** [comment]
Cool!  Look forward to seeing how you do it!

**Comment by sglanzer** [comment]
I spent some time poking at this on the weekend and it's definitely nastier than I thought going in.  The only effective strategy I've found so far is processing treeForApp on a per addon basis and capturing the output to use in coverage loader exclusions.  This approach is obviously a non-starter, but I think it clarifies what we need to make this work.

I believe the next step would be to look at the ember-cli addon build process to determine if we can peek at the broccoli trees as the addons are built and capture that output in build variable that could be used in the postBuild hook to add exclusions - basically flatten the addon app trees and exclude those patterns from the coverage.

I'll find some time to fork ember-cli and try a proof of concept, but if anyone has time to investigate this approach (or has another strategy in mind) that'd be great.

**Comment by sglanzer** [comment]
Thinking about this with a clearer head, it makes more sense to whitelist the modules in the current addon/app than to attempt to blacklist addons.  I'll explore that idea a bit - it also wouldn't require changes to ember-cli

**Comment by jschilli** [comment]
I think the issue is that there are two kinds of content that emerge from an addon - the app content and the addon content - the addon content is definitely namespaced and should be excluded.

By the time we see the code, the app content is already assimilated and there is no way that I know of to determine if it came from the consuming app or was contributed by an addon's app exports

I recall working on parts of this in ember-cli and plugin.* treeForApp blends the broccoli trees together such that post hoc determination of what came from where is not currently possible

**Comment by sglanzer** [comment]
Yeah, absolutely - that's why I think we'll need our own file structure sweep, and the whitelist appears to be the best bet currently - far less to manage than a nested addon blacklist/exclusion.

I need a solution to this soon, so I'll try to put an effort out on the weekend

**Comment by elwayman02** [comment]
:+1: looking forward to seeing some progress on this, thanks!

**Comment by sfbwalder** [comment]
:+1:  I ran into this today when I included ember-truth-helpers - suddenly many new modules under /app-name/helpers/ were included in the code coverage that had no coverage.  Filtering them out individually with either antifilter of loaderExclusions makes a long list.  

So, we would like to see this also, but we do understand it is a difficult problem since knowing what app content the add-on added isn't readily available.

**Comment by dianafa** [comment]
I ran into this a couple of days ago and it's really annoying when having a lot of addons to exclude. Is there any progress on this issue?

**Comment by flyswatter** [comment]
I doubt it, it was a pretty niche issue, I'm over it now. I believe you are the new most motivated!

- Dan
